[MCA] Adding missing instructions in AArch64 Neoverse V1 tests

jvillette38 opened the pull request:
Added missing instructions for LLVM Opcodes coverage. It will help to maintain TableGen scheduling information of AArch64 Neoverse V1.

Follow up of MR ##126703
This is a dispatch of new instructions of the big test: V1-scheduling-info.s
I have created a new test for special instructions without scheduling info in Software Optimization Guide: V1-misc-instructions.s

No more asm instruction comments to maintain.

JulienVillette committed:
[MCA] Adding missing instructions in AArch64 Neoverse V1 tests

Added missing instructions for LLVM Opcodes coverage.
It will help to maintain TableGen scheduling information of AArch64
Neoverse V1.

davemgreen commented on llvm/test/tools/llvm-mca/AArch64/Neoverse/V1-misc-instructions.s:
```diff
@@ -0,0 +1,128 @@
+dmb sy
+hlt #0x7a67
+hvc #0xecb9
+ldapur w7, [x24]
```
Atomics can go with the other atomics in basic?

davemgreen reviewed (commented):
Thanks for adding this, it looks like a nice addition to the test coverage especially the fp16 instructions. I left a number of comments looking through it.

JulienVillette committed:
[MCA] MR #128892 changes

jvillette38 commented:
> Thanks for adding this, it looks like a nice addition to the test coverage especially the fp16 instructions. I left a number of comments looking through it.

Thanks for the review! I've done changes.

davemgreen commented on llvm/test/tools/llvm-mca/AArch64/Neoverse/V1-basic-instructions.s:
```diff
@@ -19,26 +19,27 @@ add      sp, x29, #3816
 sub      w0, wsp, #4077
 sub      w4, w20, #546, lsl #12
 sub      sp, sp, #288
+sub      w13, wsp, w10
```
This one isn't add/sub immediate too.

davemgreen commented on llvm/test/tools/llvm-mca/AArch64/Neoverse/V1-basic-instructions.s:
```diff
@@ -1368,6 +1460,12 @@ ret
+# Unconditional branch (label)
+#------------------------------------------------------------------------------
+
+bl test
```
Should this be removed or is it different from the existing bl test?

davemgreen commented on llvm/test/tools/llvm-mca/AArch64/Neoverse/V1-forwarding.s:
```diff
@@ -124,6 +124,13 @@ mul    w0, w0, w0
 crc32cb w0, w0, w1
 crc32cb w0, w0, w1
 crc32cb w0, w0, w0
+crc32b w27, w12, w15
+crc32h w3, w15, w21
+crc32w w9, w18, w24
+crc32x w19, w6, x25
+crc32ch w25, w26, w16
+crc32cw w27, w12, w23
+crc32cx w21, w28, x5
```
Was that a yes or a no? :)

davemgreen commented on llvm/test/tools/llvm-mca/AArch64/Neoverse/V1-misc-instructions.s:
```diff
@@ -0,0 +1,101 @@
+isb
+pssbb
+smc #0x7e57
+stlur w3, [x27]
```
stlur can be treated as atomics too.

JulienVillette committed:
[MCA] MR #128892 changes

Removes redundant instructions in:
 AArch64/Neoverse/V1-basic-instructions.s
 AArch64/Neoverse/V1-misc-instructions.s
 AArch64/Neoverse/V1-neon-instructions.s

Fix MR remarks.

jvillette38 commented:
> > Thanks for adding this, it looks like a nice addition to the test coverage especially the fp16 instructions. I left a number of comments looking through it.
> 
> Thanks for the review! I've done changes.

Is there anything else to change? Thank you.

davemgreen commented:
It looks like there are still two unanswered questions to do with the `Unconditional branch (label)` and the `crc32b etc` additions.

jvillette38 commented:
> It looks like there are still two unanswered questions to do with the `Unconditional branch (label)` and the `crc32b etc` additions.

Sorry I pushed force this branch and lost commit with comment on crc instructions. I've pushed force back to resolve and come back again to the right commit...

JulienVillette committed:
[MCA] MR #128802 changes

Moving bl <label> to unconditionnal branch immediate.
Fixing crc32 code region name and w3 result operand.

jvillette38 commented:
> Thanks - I see part of the confusion was my fault as there isn't an existing test for bl, only for b and blr. The new test looks OK, although I would move it into the `Unconditional branch (immediate)` section (a label is just an immediate, after all).
> 
> For the new crc instructions added - did you intent to test the forwarding of crcs into one another? If not it might be better to make sure there is test coverage in V1-basic-instructions.s instead. Either way they are not the biggest deal and this looks good to me otherwise.

I moved bl <label> in unconditionnal branch (immediate section.
For crc32, other crc32 instructions were not in basic test, so I've just added in forwarding test... I did not found any dependence graph built in llvm-mca, so not easy to check forwarding between instructions:
- ``instr1 a, b, c``
- `instr2 d, a, e`

Where ``a`` should be forwarded to source operand 1 of ``instr2``. Do you confirm?

davemgreen commented:
> I moved bl in unconditionnal branch (immediate section. For crc32, other crc32 instructions were not in basic test, so I've just added in forwarding test... I did not found any dependence graph built in llvm-mca, so not easy to check forwarding between instructions:
> 
> * `instr1 a, b, c`
> * `instr2 d, a, e`
> 
> Where `a` should be forwarded to source operand 1 of `instr2`. Do you confirm?

It is shown in the pipeline diagram, which shows forwarding from crc32cb->crc32cb in the first operand, but not from mul->crc32cb and not from crc32cb->crc32cb with the second operand (or from crc32cb->mul, they all take 2 cycles as opposed to 1):
```
# CHECK:      [0,0]     DeeER.    .    ..   mul	w0, w0, w0
# CHECK-NEXT: [0,1]     D==eeER   .    ..   crc32cb	w0, w0, w1
# CHECK-NEXT: [0,2]     D===eeER  .    ..   crc32cb	w0, w0, w1
# CHECK-NEXT: [0,3]     D=====eeER.    ..   crc32cb	w0, w0, w0
# CHECK-NEXT: [1,0]     D=======eeER   ..   mul	w0, w0, w0
# CHECK-NEXT: [1,1]     D=========eeER ..   crc32cb	w0, w0, w1
# CHECK-NEXT: [1,2]     D==========eeER..   crc32cb	w0, w0, w1
# CHECK-NEXT: [1,3]     D============eeER   crc32cb	w0, w0, w0
```

It sounds like it is worth adding the instructions to V1-basic-instructions.s to make sure we have complete coverage of them all separate from where we forward.

JulienVillette committed:
[MCA] MR #128802 changes

Adding crc instructions in V1-basic-instructions.s test

davemgreen reviewed (approved):
Thanks - LGTM. (I would remove the new crc32b instructions from V1-forwarding.s too but either way it looks like a nice improvement and testing the forwarding of different crc32s has some benefit).

Are you happy for us to submit and merge?
